Update event loop documentation for gui_qt deprecation (#2639)
Napari/napari: docs/guides/connecting_events.md

(connecting-events)=

## Hooking up your own events

If you're coming from a background of scripting or working with python in an
interactive console, thinking in terms of the "event loop" can feel a bit

> strange at time.  Often we write code in a very procedural way: "do this ...

Reviewer: ```suggestion
strange at times. Often we write code in a very procedural way: "do this ...
```

then do that, etc...". With napari and other GUI programs however, usually you

> hook up a bunch of conditions and to callback functions (e.g. "If this event

Reviewer: ```suggestion
hook up a bunch of conditions to callback functions (e.g. "If this event
```

happens, then call this function") and *then* start the loop and hope you hooked
everything up correctly!  Indeed, much of the ``napari`` source code is
dedicated to creating and handling events: search the codebase for
[`.emit(`](https://github.com/napari/napari/search?q=%22.emit%28%22&type=code)
and
[`.connect(`](https://github.com/napari/napari/search?q=%22.connect%28%22&type=code)
to find examples of creating and handling internal events, respectively.

> If you would like to setup a custom event listener then you need to hook into

Reviewer: ```suggestion
If you would like to set up a custom event listener then you need to hook into
```

the napari event loop. We offer a couple of convenience decorators to easily
connect functions to key and mouse events.

### Listening for keypress events

One option is to use keybindings, that will listen for keypresses and then call
some callback whenever pressed, with the viewer instance passed as an argument
to that function. As a basic example, to add a random image to the viewer every
time the `i` key is pressed, and delete the last layer when the `k` key is
pressed:

```python
import numpy as np
import napari

viewer = napari.Viewer()

@viewer.bind_key('i')
def add_layer(viewer):
    viewer.add_image(np.random.random((512, 512)))

@viewer.bind_key('k')
def delete_layer(viewer):
    try:
        viewer.layers.pop(0)
    except IndexError:
        pass

napari.run()
```

See also this [custom key bindings
example](https://github.com/napari/napari/blob/master/examples/custom_key_bindings.py).

### Listening for mouse events

You can also listen for and react to mouse events, like a click or drag event,

> as show here where we update the image with random data every time it is

Reviewer: ```suggestion
as shown here where we update the image with random data every time it is
```

clicked.

```python
import numpy as np
import napari

viewer = napari.Viewer()
layer = viewer.add_image(np.random.random((512, 512)))

@layer.mouse_drag_callbacks.append
def update_layer(layer, event):
    layer.data = np.random.random((512, 512))

napari.run()
```

See also the [custom mouse
functions](https://github.com/napari/napari/blob/master/examples/custom_mouse_functions.py)
and [mouse drag
callback](https://github.com/napari/napari/blob/master/examples/mouse_drag_callback.py)
examples.

### Connection functions to native napari events

If you want something to happen following some event that happens *within*

> napari, then trick becomes knowing which native signals any given napari object

Reviewer: ```suggestion
napari, the trick becomes knowing which native signals any given napari object
```
